ddtrace/profiling/collector/_memalloc.c

```diff
@@ -394,20 +394,18 @@ iterevents_new(PyTypeObject* type, PyObject* Py_UNUSED(args), PyObject* Py_UNUSE
     }
 
     IterEventsState* iestate = (IterEventsState*)type->tp_alloc(type, 0);
-    if (!iestate)
+    if (!iestate) {
+        PyErr_SetString(PyExc_RuntimeError, "failed to allocate IterEventsState");
         return NULL;
+    }
 
     memalloc_assert_gil();
 
     /* reset the current traceback list */
-    if (memlock_trylock(&g_memalloc_lock)) {
-        iestate->alloc_tracker = global_alloc_tracker;
-        global_alloc_tracker = alloc_tracker_new();
-        memlock_unlock(&g_memalloc_lock);
-    } else {
-        Py_TYPE(iestate)->tp_free(iestate);
-        return NULL;
-    }
+    memlock_lock(&g_memalloc_lock);
```

Comment:
Since we're changing from `try_lock` to `lock`, would this be safe from indefinitely blocking?

Reply:
Good idea. I actually ended up pulling the allocation of the new tracker outside the lock in https://github.com/DataDog/dd-trace-py/pull/12075/commits/6c682abdf1e819c5dcb6c19d00d309b70c3bcefe, and added a comment for why. Please take another look 🙏